### ISSUE TITLE
use prototypal inheritance

### ISSUE THREAD

**Issue by derhuerst** [brianshaler/hifo-stream#1]
The reason I use [prototypal inheritance](http://davidwalsh.name/javascript-objects-deconstruction#simpler-object-object) is that you can easily customize the `Hifo` "class".

So this a more elegant way:

``` javascript
var hifo = require('hifo');

// modify class
hifo.Hifo.add = …

// now use it
var h = hifo(…);
```

**Comment by derhuerst** [comment]
Because the `require`d _hifo_ module gets cached, this changes the `Hifo` "class" everywhere it's `require`d. To avoid this, create a "subclass":

``` javascript
var hifo = require('hifo');

// create a subclass
MyHifo = Object.create(hifo.Hifo);
// modify class
MyHifo.add = …

// now use it
var h = Object.create(MyHifo);
h.init(…);
```

**Comment by brianshaler** [comment]
Are you saying hifo-stream should be written that way in order to expose an extensible base class? Or that it should extend hifo and replace its .add() method?

**Comment by derhuerst** [comment]
@brianshaler 

The latter. Even though it's ridiculous at the moment, theoretically you don't leverage one of the benefits of modularization now. If I fixed a bug in _hifo_, you would, extending _hifo_, also benefit from it.

So my proposal is to let _hifo-stream_ extend _hifo_.

**Comment by brianshaler** [comment]
That's where I started, but I ended up removing the dependency on _hifo_ because most of _hifo_'s logic is within `.add()`. If there was a way I could know if `.add()` resulted in a change to `data`, I would use _hifo_'s `.add()` as well as its `.lowest()` and `.highest()` helper functions.

**Comment by derhuerst** [comment]
> That's where I started, but I ended up removing the dependency on hifo because most of hifo's logic is within .add().

That's a valid point. :D

> If there was a way I could know if .add() resulted in a change to data, I would use hifo's .add() as well as its .lowest() and .highest() helper functions.

Feel free to propose changes to _hifo_, as long as it stays a usable module and isn't just a helper for _hifo-stream_.

**Comment by derhuerst** [comment]
I accepted your pull request derhuerst/hifo#3 and released [version `0.4.0`](https://github.com/derhuerst/hifo/releases/tag/0.4.0).

**Comment by brianshaler** [comment]
@derhuerst 

`.insert()` seems to have done the trick. Thanks!

[./index.js](https://github.com/brianshaler/hifo-stream/blob/master/index.js) is now half as many LOC and uses all of _hifo_'s functionality and doesn't need to re-implement any of it.
